engine/kv_engine_cleaner.cpp

```diff
@@ -617,22 +605,177 @@ void Cleaner::AdjustThread(size_t advice_thread_num) {
   }
 }
 
+double Cleaner::SearchOutdatedCollections() {
+  size_t limited_fetch_num = 0;
+  std::unique_lock<SpinMutex> queue_lock(outdated_collections_.queue_mtx_);
+  int64_t before_queue_size =
+      static_cast<int64_t>(outdated_collections_.hashlists.size() +
+                           outdated_collections_.lists.size() +
+                           outdated_collections_.skiplists.size());
+  {
+    std::unique_lock<std::mutex> skiplist_lock(kv_engine_->skiplists_mu_);
+    auto iter = kv_engine_->expirable_skiplists_.begin();
+    while (!kv_engine_->expirable_skiplists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto outdated_skiplist = *iter;
+      iter = kv_engine_->expirable_skiplists_.erase(iter);
+      outdated_collections_.skiplists.emplace(outdated_skiplist);
+      limited_fetch_num++;
+    }
+  }
+
+  {
+    std::unique_lock<std::mutex> list_lock(kv_engine_->lists_mu_);
+    auto iter = kv_engine_->lists_.begin();
+    while (!kv_engine_->lists_.empty() && (*iter)->HasExpired() &&
+           limited_fetch_num < max_thread_num_) {
+      auto expired_list = *iter;
```

Comment:
Only latest version of list is present on the HashTable/lists_. Older versions are chained after newer versions. The std::set actually no longer makes sense as older versions are no longer in it and the scan must scan the whole chain to find the outdated versions.
If all lists are deleted once and then immediately created again, this scan can find nothing outdated.